### PR TITLE
Brownian bridge

### PR DESCRIPTION
**Pull request by fabrice-deschatres**
Adding methods for single and double brownian bridges.

### CODE REVIEW

**Comment by fabrice-deschatres** [comment]
@googlebot I signed it!

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+def brownian_bridge_double(*,
+                           x_start,
+                           x_end,
+                           variance,
+                           upper_barrier,
+                           lower_barrier,
+                           n_cutoff=3,
+                           dtype=None,
+                           name=None):
+  """ Probability of not touching the upper barrier and lower_barrier
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: The first line of docstring should be a sentence < 80 symbols followed by a whitespace

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  bi-dimensional (with time steps and paths being the 2 dimensions).
+
+  ## References
+
+  [1] Emmanuel Gobet. Advanced Monte Carlo methods for barrier and related
+  exotic options.
+  https://papers.ssrn.com/sol3/papers.cfm?abstract_id=1265669
+
+  Args:
+    x_start: Real `Tensor` of any shape and dtype
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: `A real` for consistency

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                           name=None):
+  """ Probability of not touching the upper barrier and lower_barrier
+  for a Brownian Bridge starting at `x_start`, ending at `x_end` and with
+  variance `variance`, assuming that `x_start` and `x_end` are within the
+  barriers.
+  This can be used in Monte Carlo pricing for adjusting probability of
+  touching the barriers from discrete case to continuous case.
+  Typically in practise, the tensors `x_start`, `x_end` and `variance` should be
+  bi-dimensional (with time steps and paths being the 2 dimensions).
+
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Could you please add an example? See, e.g., [here](
https://github.com/google/tf-quant-finance/blob/ea804648b9461d32675c63844ccae8c4da9f48e3/tf_quant_finance/black_scholes/vanilla_prices.py#L37)

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    lower_barrier: A scalar `Tensor`
+    n_cutoff: A positive scalar int32 `Tensor`. This controls when to cutoff
+      the sum which would otherwise have an infinite number of terms.
+      Default value: 3.
+    dtype: Optional `tf.DType`. If supplied, the dtype to be used for conversion
+      of any supplied non-`Tensor` arguments to `Tensor`.
+      Default value: None which maps to the default dtype inferred by TensorFlow
+        (float32).
+    name: str. The name for the ops created by this function.
+      Default value: None which is mapped to the default name `option_price`.
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Pleas change the default name to `brownian_bridge_double `

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+  Returns:
+      A `Tensor` of the same shape as the input data which is the probability
+      of not touching the barrier.
+  """
+  with tf.name_scope(name or 'single_brownian_bridge'):
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: The default name should be the same as the function name unless there is a specific reason not to do so. Should we use `brownian_bridge_single `?

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    x_start = tf.convert_to_tensor(x_start, dtype=dtype, name='x_start')
+    dtype = x_start.dtype
+    variance = tf.convert_to_tensor(variance, dtype=dtype, name='variance')
+    x_end = tf.convert_to_tensor(x_end, dtype=dtype, name='x_end')
+    barrier_diff = upper_barrier - lower_barrier
+
+    def summed_func(k):
+      return tf.exp(-2 * k * barrier_diff * (
+              k * barrier_diff + (x_end - x_start)) / variance) \
+             - tf.exp(-2 * (k * barrier_diff + x_start - upper_barrier) * (
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Please don't use `\` to split the lines as this violates the style guide

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+  """
+  with tf.name_scope(name or 'double_brownian_bridge'):
+    x_start = tf.convert_to_tensor(x_start, dtype=dtype, name='x_start')
+    dtype = x_start.dtype
+    variance = tf.convert_to_tensor(variance, dtype=dtype, name='variance')
+    x_end = tf.convert_to_tensor(x_end, dtype=dtype, name='x_end')
+    barrier_diff = upper_barrier - lower_barrier
+
+    def summed_func(k):
+      return tf.exp(-2 * k * barrier_diff * (
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: `tf.math.exp` for consistency (since, e.g., tf.log is not exposed and we use tf.math.log instead)

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+              k * barrier_diff + (x_end - x_start)) / variance) \
+             - tf.exp(-2 * (k * barrier_diff + x_start - upper_barrier) * (
+              k * barrier_diff + (x_end - upper_barrier)) / variance)
+
+    return tf.reduce_sum(
+      tf.concat(
+        tf.map_fn(summed_func, tf.range(-n_cutoff, n_cutoff + 1, dtype=dtype)),
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: In my experience `tf.map_fn` is extremely slow.

In this case, you should be able to utilize vectorization of tf functions and write something like

tf.reduce_sum(summed_func(tf.range(-n_cutoff, n_cutoff + 1, dtype=dtype)), axis=0)

You might need to do some broadcasting like tf.expand_dims(tf.range(-n_cutoff, n_cutoff + 1, dtype=dtype), 0).

Let me know if you need help with this one

**Comment by fabrice-deschatres** [comment]
Thanks for the feedback Cyril.

Regarding the last point, this works without map_fn:

with tf.name_scope(name or 'brownian_bridge_double'):
  x_start = tf.convert_to_tensor(x_start, dtype=dtype, name='x_start')
  dtype = x_start.dtype
  variance = tf.convert_to_tensor(variance, dtype=dtype, name='variance')
  x_end = tf.convert_to_tensor(x_end, dtype=dtype, name='x_end')
  barrier_diff = upper_barrier - lower_barrier

  original_shape = x_start.shape

  x_start = tf.expand_dims(tf.reshape(x_start,[tf.size(x_start)]),1)
  x_end = tf.expand_dims(tf.reshape(x_end,[tf.size(x_end)]),1)
  variance = tf.expand_dims(tf.reshape(variance,[tf.size(variance)]),1)

  k = tf.expand_dims(tf.range(-n_cutoff, n_cutoff + 1, dtype=dtype), 0)

  a = k * barrier_diff * (k * barrier_diff + (x_end - x_start))
  b = (k * barrier_diff + x_start - upper_barrier)
  b *= k * barrier_diff + (x_end - upper_barrier)

  output = tf.math.exp(- 2 * a / variance) - tf.math.exp(-2 * b / variance)

  return tf.reshape(tf.reduce_sum(output, axis=1), original_shape)


But the drawback is that it flattens the input array (and reshape at the
end) which (1) looks awkward, and (2) doesn't actually works if x_start,
x_end and variance don't have all the same shape (which is the case in the
unittest as variance is 1D vs x_start and x_end which are 2D). Could you
think of a better way?

**Comment by cyrilchim** [comment]
Hi Fabrice, 

You don't need to flatten the tensors as broadcasting will happen automatically. You just need to expand the last dimension. The code below should work

```
  x_start = tf.expand_dims(x_start, -1)
  x_end = tf.expand_dims(x_end, - 1)
  variance = tf.expand_dims(variance ,-1)

  k = tf.expand_dims(tf.range(-n_cutoff, n_cutoff + 1, dtype=dtype), 0)

  a = k * barrier_diff * (k * barrier_diff + (x_end - x_start))
  b = (k * barrier_diff + x_start - upper_barrier)
  b *= k * barrier_diff + (x_end - upper_barrier)

  output = tf.math.exp(- 2 * a / variance) - tf.math.exp(-2 * b / variance)

  return tf.reduce_sum(output, axis=-1)
```

**Comment by fabrice-deschatres** [comment]
Thanks Cyril, that worked. I've updated my fork, I believe all your comments have been addressed.

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Methods for brownian bridges.
+
+These can be used in Monte-Carlo simulation for payoff with continuous barrier.
+Indeed, the Monte-Carlo simulation is inherently discrete in time, and to
+improve convergence (w.r.t. the number of time steps) for payoff with continuous
+barrier, adjustment with brownian bridge can be made.
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Could you please add a reference here too?

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+                           x_end,
+                           variance,
+                           upper_barrier,
+                           lower_barrier,
+                           n_cutoff=3,
+                           dtype=None,
+                           name=None):
+  """ Probability of not touching any of the barriers
+  (`upper_barrier` and `lower_barrier`) for a Brownian Bridge starting at
+  `x_start`, ending at `x_end` and with variance `variance`, assuming that
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Just a nitpick but the first sentence in the docstring should be an imperative verb with the first line being <=80 symbols ending with a period and separated by a blank line. Say, here,  

"""Computes probabilities of not touching the boundaries  for a Brownian Bridge.

 For a Brownian Bridge starting at `x_start`, ending at `x_end` and with the specified `variance` ... blah blah

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+
+def brownian_bridge_single(*,
+                           x_start,
+                           x_end,
+                           variance,
+                           barrier,
+                           dtype=None,
+                           name=None):
+  """ Probability of not touching the barrier for a Brownian Bridge starting
+  at `x_start`, ending at `x_end` and with variance `variance`,
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Same as above

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+  exotic options.
+  https://papers.ssrn.com/sol3/papers.cfm?abstract_id=1265669
+
+  Args:
+    x_start: A real `Tensor` of any shape and dtype
+    x_end: A real `Tensor` of the same dtype and compatible shape as
+      `x_start`
+    variance: A real `Tensor` of the same dtype and compatible shape as
+      `x_start`
+    barrier: A scalar `Tensor`
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Please add a period at the end of each arg definition. Also, please expand on the meaning of barrier

barrier: A scalar `Tensor` of the same `dtype` as `x_start`. Stands for the boundary for the Brownian Bridge.

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+    dtype: Optional `tf.DType`. If supplied, the dtype to be used for conversion
+      of any supplied non-`Tensor` arguments to `Tensor`.
+      Default value: None which maps to the default dtype inferred by TensorFlow
+        (float32).
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Just 
Default value: None which maps to the default dtype inferred by TensorFlow. 

No need for float32 as the default value might be as well float64

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+    lower_barrier: A scalar `Tensor`
+    n_cutoff: A positive scalar int32 `Tensor`. This controls when to cutoff
+      the sum which would otherwise have an infinite number of terms.
+      Default value: 3.
+    dtype: Optional `tf.DType`. If supplied, the dtype to be used for conversion
+      of any supplied non-`Tensor` arguments to `Tensor`.
+      Default value: None which maps to the default dtype inferred by TensorFlow
+        (float32).
+    name: str. The name for the ops created by this function.
+      Default value: None which is mapped to the default name `option_price`.
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Please change the default name to `brownian_bridge_double`

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+  Args:
+    x_start: A real `Tensor` of any shape and dtype
+    x_end: A real `Tensor` of the same dtype and compatible shape as
+      `x_start`
+    variance: A real `Tensor` of the same dtype and compatible shape as
+      `x_start`
+    barrier: A scalar `Tensor`
+    dtype: Optional `tf.DType`. If supplied, the dtype to be used for conversion
+      of any supplied non-`Tensor` arguments to `Tensor`.
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Please copy the description from above

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+      Default value: None which maps to the default dtype inferred by TensorFlow
+        (float32).
+    name: str. The name for the ops created by this function.
+      Default value: None which is mapped to the default name `option_price`.
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Default name is `brownian_bridge_single`

**Comment by fabrice-deschatres** [comment]
OK, I've updated BUILD accordingly.

**Comment by cyrilchim** [comment]
Fabrice, could you please also resolve the other comments I've left? Thanks!

**Comment by fabrice-deschatres** [comment]
OK, sorry, I missed these comments. I've updated again.

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+def brownian_bridge_double(*,
+                           x_start,
+                           x_end,
+                           variance,
+                           upper_barrier,
+                           lower_barrier,
+                           n_cutoff=3,
+                           dtype=None,
+                           name=None):
+  """ Computes probabilities of not touching the barriers for a Brownian
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Could you please update the docstring as described [here](http://google.github.io/styleguide/pyguide.html#381-docstrings)? Something like
"""Computes probabilities of not touching the barriers for a Brownian Bridge.

The Brownian bridge starts at ... 
"""

There needs to be a blank line between the first line and the rest
Also, the docstring for `brownian_bridge_single` needs to be updated

File: tf_quant_finance/black_scholes/brownian_bridge.py [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+  Bridge starting at `x_start`, ending at `x_end` and with variance `variance`,
+  assuming that `x_start` and `x_end` are within the barriers 'lower_barrier'
+  and 'upper_barrier'.
+  This can be used in Monte Carlo pricing for adjusting probability of
+  touching the barriers from discrete case to continuous case.
+  Typically in practise, the tensors `x_start`, `x_end` and `variance` should be
+  bi-dimensional (with time steps and paths being the 2 dimensions).
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/brownian_bridge.py]: Just to clarify, is this formula is for 1-dimensional Brownian Bridge? If so, could you please say so explicitly in the docstring?

**Comment by fabrice-deschatres** [comment]
Yes, 1D brownian bridge. I have adjusted the docstring, let me know if further comments.

**Comment by cyrilchim** [comment]
Thank you, Fabrice! This is now checked in

**Comment by fabrice-deschatres** [comment]
Great, thanks for reviewing Cyril!
